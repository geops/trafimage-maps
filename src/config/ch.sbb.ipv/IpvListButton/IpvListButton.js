import React, { useEffect, useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { unByKey } from 'ol/Observable';
import { extend, createEmpty, getCenter } from 'ol/extent';
import { Point } from 'ol/geom';
import MapButton from '../../../components/MapButton';
import { setFeatureInfo } from '../../../model/app/actions';
import { IPV_TOPIC_KEY } from '../../../utils/constants';
import { ReactComponent as List } from '../../../img/list-icon-sbb.svg';

const getExtentFromFeatures = (features) => {
  let extent = createEmpty();
  features.forEach((feat) => {
    extent = extend(extent, feat.getGeometry().getExtent());
  });
  return extent;
};

const IpvListButton = () => {
  const dispatch = useDispatch();
  const { t } = useTranslation();
  const layers = useSelector((state) => state.map.layers);
  const map = useSelector((state) => state.app.map);
  const featureInfo = useSelector((state) => state.app.featureInfo);
  const ipvMainLayer = useMemo(
    () => layers.find((l) => l.key === `${IPV_TOPIC_KEY}.mainlayer`),
    [layers],
  );
  const [features, setFeatures] = useState([]);

  useEffect(() => {
    const loadFeatsListener = ipvMainLayer?.on('load:features', (evt) => {
      setFeatures(evt.features);
    });
    return () => unByKey(loadFeatsListener);
  }, [ipvMainLayer, map, featureInfo, dispatch, features]);

  return (
    <MapButton
      title={t('Alle Direktverbindungen anzeigen')}
      disabled={!features?.length}
      style={{ padding: 8 }}
      onClick={() => {
        if (!features?.length) return;
        const view = map.getView();
        const extent = getExtentFromFeatures(features);
        view.cancelAnimations();
        view.fit(new Point(getCenter(extent)), {
          duration: 500,
          padding: [100, 100, 100, 100],
          maxZoom: 6,
          callback: () => {
            const { mbMap } = ipvMainLayer.mapboxLayer;
            if (mbMap) {
              mbMap.once('idle', () => {
                dispatch(
                  setFeatureInfo([
                    {
                      features: ipvMainLayer.syncFeatures(),
                      layer: ipvMainLayer,
                    },
                  ]),
                );
              });
            }
          },
        });
      }}
    >
      <List />
    </MapButton>
  );
};

IpvListButton.propTypes = {};

export default IpvListButton;
