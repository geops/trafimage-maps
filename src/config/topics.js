import defaultBaseLayers, {
  bahnhofplaene,
  passagierfrequenzen,
  netzkarteLayer,
  netzkartePointLayer,
  buslines,
  gemeindegrenzen,
  punctuality,
  netzkarteShowcases,
  parks,
  stuetzpunktbahnhoefe,
} from './layers';
import defaultSearches, { handicapStopFinder } from './searches';

const defaultElements = {
  header: true,
  footer: true,
  menu: true,
  permalink: true,
  mapControls: true,
  baseLayerToggler: true,
  popup: false,
  search: true,
};

export const netzkarte = {
  name: 'ch.sbb.netzkarte',
  key: 'ch.sbb.netzkarte',
  elements: {
    ...defaultElements,
    popup: true,
    shareMenu: true,
    trackerMenu: true,
  },
  layers: [
    ...defaultBaseLayers,
    gemeindegrenzen,
    parks,
    punctuality,
    buslines,
    netzkartePointLayer,
    passagierfrequenzen,
    bahnhofplaene,
  ],
  projection: 'EPSG:3857',
  description: 'ch.sbb.netzkarte-desc',
  searches: defaultSearches,
};

export const handicap = {
  name: 'ch.sbb.handicap',
  key: 'ch.sbb.handicap',
  elements: { ...defaultElements, shareMenu: true, popup: true },
  layers: [...defaultBaseLayers, stuetzpunktbahnhoefe],
  projection: 'EPSG:3857',
  description: 'ch.sbb.handicap-desc',
  searches: { Stationen: handicapStopFinder },
};

export const netzkarteStelen = {
  name: 'ch.sbb.netzkarte',
  key: 'ch.sbb.netzkarte',
  layers: [netzkarteLayer],
  projection: 'EPSG:3857',
};

export const casa = {
  name: 'CASA',
  key: 'ch.sbb.casa',
  layers: [netzkarteLayer],
  projection: 'EPSG:3857',
};

export const bauprojekte = {
  name: 'ch.sbb.construction',
  key: 'ch.sbb.construction',
  linkUrl: 'https://maps.trafimage.ch/#/ch.sbb.construction',
  description: 'ch.sbb.construction-desc',
};

export const behig = {
  name: 'ch.sbb.behig',
  key: 'ch.sbb.behig',
  linkUrl: 'https://maps.trafimage.ch/#/ch.sbb.infrastruktur',
  description: 'ch.sbb.behig-desc',
  legendUrl: '/img/topics/behig/behig_legend_{language}.jpg',
};

export const infrastruktur = {
  name: 'ch.sbb.infrastruktur',
  key: 'ch.sbb.infrastruktur',
  linkUrl: 'https://maps.trafimage.ch/#/ch.sbb.infrastruktur',
  description: 'ch.sbb.infrastruktur-desc',
};

export const regionenkarte = {
  name: 'ch.sbb.regionenkarte.public',
  key: 'ch.sbb.regionenkarte.public',
  linkUrl: 'https://maps.trafimage.ch/#/ch.sbb.regionenkarte.public',
  description: 'ch.sbb.regionenkarte.public-desc',
};

export const tarifverbundkarte = {
  name: 'ch.sbb.tarifverbundkarte.public',
  key: 'ch.sbb.tarifverbundkarte.public',
  linkUrl: 'https://maps.trafimage.ch/#/ch.sbb.tarifverbundkarte.public',
  description: 'ch.sbb.tarifverbundkarte.public-desc',
  legendUrl: '/img/topics/tarifverbundkarte/tarifverbundkarte_legend.png',
};

export const showcases = {
  name: 'ch.sbb.showcases',
  key: 'ch.sbb.showcases',
  elements: {
    ...defaultElements,
    baseLayerToggler: false,
  },
  layers: [netzkarteShowcases],
  projection: 'EPSG:3857',
  description: 'ch.sbb.showcases-desc',
};

export default [
  netzkarte,
  handicap,
  bauprojekte,
  behig,
  infrastruktur,
  regionenkarte,
  tarifverbundkarte,
  showcases,
];
