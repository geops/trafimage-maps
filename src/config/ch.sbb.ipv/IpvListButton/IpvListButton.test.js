import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import OLMap from 'ol/Map';
import ipvLayers from '../index';
import { ipvIframe } from '../../topics';
import IpvListButton from './IpvListButton';

describe('IpvListButton', () => {
  const mockStore = configureStore([thunk]);
  let store;

  beforeEach(() => {
    store = mockStore({
      map: { layers: ipvLayers },
      app: {
        map: new OLMap({}),
        activeTopic: ipvIframe,
        featureInfo: [],
        displayMenu: true,
      },
    });
  });

  test('should match snapshot and be disabled on load.', () => {
    const { container, getByTestId } = render(
      <Provider store={store}>
        <IpvListButton />
      </Provider>,
    );
    expect(container.innerHTML).toMatchSnapshot();
    expect(getByTestId('ipv-list-button')).toBeDisabled();
  });
});
