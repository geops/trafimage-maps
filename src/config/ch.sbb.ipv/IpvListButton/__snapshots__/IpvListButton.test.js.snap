// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IpvListButton should match snapshot and be disabled on load. 1`] = `
<button class="MuiButtonBase-root MuiIconButton-root WithStyles(ForwardRef(IconButton))-root-1 Mui-disabled Mui-disabled"
        tabindex="-1"
        type="button"
        disabled
        title="Alle Direktverbindungen anzeigen"
        data-testid="ipv-list-button"
>
  <span class="MuiIconButton-label">
    <svg>
      list-icon-sbb.svg
    </svg>
  </span>
</button>
`;
